README
HTTP Replicator
===============
Gertjan van Zwieten <gertjanvanzwieten@fastmail.fm>

:Project ID: 44954

.http://freshmeat.net/projects/http-replicator[Freshmeat Project]
**********************************************************************
HTTP Replicator is a general purpose, replicating HTTP proxy server.
All downloads through the proxy are checked against a private cache,
which is an exact copy of the remote file structure. If the requested
file is in the cache, replicator sends it out at LAN speeds. If not in
the cache, it will simultaneously download the file and stream it to
multiple clients. No matter how many machines request the same file,
only one copy comes down the Internet pipe. Very useful for
maintaining a cache of Debian or Gentoo packages.
**********************************************************************


History
-------

A while ago, I was looking for a way to cache my downloaded
http://www.debian.org[debian] (unstable) packages. I have some
computers on a LAN, and it seemed like a waste of time and bandwidth
to download the same packages over and over again for each computer.
So I looked into some existing solutions... 

http://www.squid-cache.org[squid]
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Squid is a very well known proxy server, which (amongst others)
maintains a cache of downloaded files. Quite a few people seem to use
squid for the purposes I'm interested in. What bothers me is that
these files are accessible only by squid itself. It is (as far as I
know) not possible to get a list of all cached files, let alone to
manually install or copy a file from cache. This is probably a
consequence of squids immense functionality, but for my purpose it
seems most logical to just store files as files. 

http://apt-proxy.sourceforge.net[apt-proxy]
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Apt-proxy is a proxy server written specially for the purpose of
maintaining a debian cache. The current version is a bash script, so
speed is not optimal but it easily beats my internet connection...
Actually it isn't a real proxy server but a normal http server that
serves files from some configured remote hosts.  These hosts are
configured in both sources.list and the apt-proxy configuration file.
I believe either location has its pros and cons, but in requiring both
you'll end up with cons only. I've used it for a while, but I got a
strange http error every now and then that screwed up apt so I had to
constantly restart.  There is also a beta rewrite in
http://www.python.org[python], which I also tried, but I don't
remember being to satisfied with that either. 

http://www.apt-cacher.org[apt-cacher]
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
This one looked really promising. It is a specialized debian proxy
server, just like apt-proxy, but it is written in perl as an apache
module. Again, this isn't a real proxy, but this time sources are
defined in sources.list only so that's an improvement. Moreover, this
one works (!) and it even seems to store files in the normal way that
seems so obvious to me. But... when inspecting the files in cache they
appear to be somehow altered, for a reason I cannot understand, but
they can't be viewed nor installed manually. Apart from that I don't
like the idea of running an entire apache server for a debian cache
only. 

http://sourceforge.net/projects/debproxy[debproxy]
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Yet another specialized proxy server, a true proxy this time. As you
can see from the "this project has not released any files" it never
came out of cvs. I can't remember if I ever had it up and running.
However judging from the description this project would have been
ideal if it would have been finished.  It is a real proxy, simply
caching all files that are downloaded through it or serving from cache
if they're there already. Creator
http://sourceforge.net/users/lordsutch[lordsuch] named this kind of
proxy a "replicating proxy". This name seemed so obvious that I
immediately googled for more... but only debproxy showed up. That's
when I began to realize that the proxy I was looking for didn't yet
exist. Hence my own attempt: 


Replicator
----------

Replicator is a general purpose proxy server. That is, not
specifically geared towards debian caching. It can be used for any
http client, for instance a web browser. While browsing the web it
will simply save all the files that go through it. The data is
transparently streamed to all clients simultaneously requesting the
same file, so everything is downloaded only once. When a requested
file is already in cache a check is added to the http header to see if
the file is modified since the time it was cached. If it isn't than
the file is served from cache, preventing it from being downloaded
again. Real speed gains are obtained when the server is configured not
to check for modifications but to serve the cached file directly. 

To use replicator for debian caching modify or create
/etc/apt/apt.conf to contain the following line: 

--------------------------------------------
Acquire::http::Proxy "http://(HOST):(PORT)"; 
--------------------------------------------

Now apt and similar will download through the replicator proxy,
caching all packages that are installed. This has worked flawlessly on
my computer for quite some time now. The advantage of a general proxy
rather than a specialized one becomes apparent when installing
packages such as 'msttcorefonts' and 'flashplugin-nonfree'. These
packages offer the possibility to use a proxy server, so not only the
packages themselves but also related files can be cached. As of
version 2.0 replicator is also suitable for maintaining a
http://www.gentoo.org[gentoo] package cache, mainly because of the new
flat mode - see below. And as of version 2.1 it is also possible to
forward requests to an external proxy server. 

In version 3.0 a lot of new features are introduced, such as... cache
browsing!  Yes, that's right, replicator can now act as a web server.
Simply visit the address where you have replicator configured to
listen for proxy connections in a browser and you'll get an html
listing of the files in cache. And the great thing is that the
streaming principle still applies, so even incomplete files can be
served. Another new feature is client-side support for range requests,
so incomplete downloads can now be resumed. For a complete list of
changes see the link:changelog/[changelog]. All this added
functionality required some fairly 'deep' changes in code. The
interested will notice that the `LocalResponse` construct has
disappeared, which is what I personally like most about this version
:-). 

The following command line options control the server: 

Options
~~~~~~~
-p, --port PORT::
  The proxy port on which the server listens for http requests.
  Default is 8080. 

-i, --ip IP::
  The ip addresses from which access is allowed, optionally with
  wildcards ? and \* for single and multiple digits respectivily.
  Defaults to localhost * and can be used multiple times. 

-d, --dir DIR::
  The directory to store the cached files in. By default files are
  simply saved under the current directory. 

-a, --alias STR::
  Colon separated list of mirrors, i.e. hosts serving identical
  content. Useful for caching (debian) packages from different mirrors
  to the same directory, like this:
  'cache/dir:host1/somepath:host2/other/path'. Can be used multiple
  times. 

-s, --static::
  Static mode: files are known to never change so files that are
  present are served from cache directly without contacting the
  server. Primarily useful for gentoo caching, not for debian because
  of the changing Packages.gz and Release files. 

-f, --flat::
  Flat mode: all files are saved in a single directory. Primary use
  for gentoo. 

-e, --external::
  Forward requests to an external proxy server, specified as host:port
  or username:password@host:port if the server requires
  authentication. 

-q, --quiet::
  Decrease verbosity. Messages on stdout all have a certain level of
  importancy.  By repeatedly selecting this option the least important
  messages are skipped. 

Daemon options
~~~~~~~~~~~~~~
--daemon::
  Switches to daemon mode and activates the following options: 

--log LOG::
  Write output to a log file instead of stdout. 

--pid PID::
  Write the process id to a pid file instead of stdout. 

--user USER::
  Change the process user id after opening the log and pid file. 

Debugging options
~~~~~~~~~~~~~~~~~
--debug::
  Set the lowest possible logging level (the quiet option is ignored)
  and activate the following options: 

--intolerant::
  Crash on unhandled exceptions. 

When the server is started from `/etc/init.d` the command line options
are read from `/etc/default/http-replicator`. Here two variables are
defined: `DAEMON_OPTS` that contains the command line options for the
server and `MAINTENANCE_OPTS` that contains the options for the
maintenance (cron) script.  When replicator is just installed all
files are in place but the server is not yet activated. For that you
must remove the 'exit' command in `/etc/default/http-replicator`. This
is done for two reasons. First, you might not want to have the server
running all the time but just use it from the command line. As I said,
package caching is not replicator's only purpose.  Second, if you do
want to have replicator running all the time you need to create the
caching directory first and set up the right permissions. You may also
want to change the port and select the static or flat operation mode
before starting the server. 

I already mentioned the cron maintenance script that comes with
replicator. This script helps keeping the package cache down in size.
Without such script a package cache tends to grow very rapidly to
immense proportions. The maintenance script simply deletes all
packages of which more than a certain number of newer versions are
present. This number is set on the commandline with `--keep`, which is
read from `/etc/default/http-replicator`. The script is disabled by
setting the number to zero or less. 

Links
-----

Some websites that are somehow related to this project: 

* http://gertjan.freezope.org/replicator[replicator home]
* http://freshmeat.net/projects/http-replicator[freshmeat project
  page]
* http://gentoo-wiki.com/HOWTO_Download_Cache_for_LAN-Http-Replicator[gentoo
  specific replicator howto]
* http://forums.gentoo.org/viewtopic.php?t=173226[gentoo discussion
  forum with replicator ebuild]
* http://www.xhaus.com/alan/python/proxies.html[overview of http
  proxies written in python]
* http://www.jmarshall.com/easy/http[nice introduction to the HTTP
  protocol]
* http://www.w3.org/Protocols/rfc2616/rfc2616.html[complete HTTP/1.1
  protocol]


ChangeLog
xxxx-xx-xx
  * changed umask to 022
      (Corey Wright: umask.patch)
  * relaxed timefmt constraints in http header
      (Corey Wright: multiple_last-modified_header_formats.patch)
  * fixed long filename issue by hashing excess length

2008-31-01 version 4.0alpha2
  * added GPL licence file
  * generalized fiber.py by using generator name instead of hardcoded string
  * removing partial file in cache after 403 forbidden
  * flushing every line in debug mode
  * fix: no longer setting mtime to -1 if server does not send Last-Modified
  * fix: handling empty command lines correctly

2008-01-01 version 4.0alpha1
  * rewrite from scratch
  * replaced restrictive asyncore scheduler with new 'fiber' system
  * new feature: server-side download resuming
  * new feature: fpt support
  * new feature: bandwidth shaping
  * new feature: ipv6 support
  * new feature: frozen transactions killed after configurable timeout
  * new feature: rudimentary off-line mode
  * fixed race condition that prevented joining of simultaneously started downloads
  * currently missing feature: cache browsing

2004-11-27 version 3.0
  * new feature: cache brower on proxy address
  * new feature: client-side support for partial content
  * added alias option for caching mirrors on same location
  * added check to prevent access outside of cache through symlinks
  * added header length restriction to fight infinite request server attacks
  * created man pages for http-replicator and http-replicator_maintenance
  * fixed timestamp bug; files are now properly closed before changing mtime
  * suppressed size warning for chunked data

2004-08-15 version 2.1
  * integrated daemon code in http-replicator
  * changed init.d and cron script to bash
  * moved settings from configuration file to /etc/default/http-replicator
  * introduced optparse module for command line parsing
  * introduced logging module for output
  * added support for an external proxy server
  * added support for an external proxy requiring authentication

2004-05-01 version 2.0
  * added support for HTTP/1.1
  * replicator is now suitable for maintaining a gentoo package cache
  * fixed problem with absolute urls
  * added posting support
  * added support for servers that use LF in header instead of CRLF
  * added a command line system
  * fixed security issues
  * improved traceback message for unhandled exceptions
  * fixed problem with incomplete files after a server crash
  * fixed problems with select
  * fixed size calculation in cron script

2004-02-06 version 1.0
  * initial release.


Params.py
import sys, os, socket

_args = iter( sys.argv )

PROG = _args.next()
PORT = 8080
ROOT = os.getcwd() + os.sep
VERBOSE = 0
TIMEOUT = 15
FAMILY = socket.AF_INET
FLAT = False
STATIC = False
ONLINE = True
LIMIT = False
LOG = False
DEBUG = False
MAXCHUNK = 1448 # maximum lan packet?
TIMEFMT = ('%a, %d %b %Y %H:%M:%S GMT', '%a, %d %b %Y %H:%M:%S +0000 GMT', '%a, %d %b %Y %H:%M:%S +0000')
SUFFIX = '.incomplete'
USAGE = '''usage: %(PROG)s [options]

options:
  -h --help          show this help message and exit
  -p --port PORT     listen on this port for incoming connections, default %(PORT)i
  -r --root DIR      set cache root directory, default current: %(ROOT)s
  -v --verbose       show http headers and other info
  -t --timeout SEC   break connection after so many seconds of inactivity, default %(TIMEOUT)i
  -6 --ipv6          try ipv6 addresses if available
     --flat          flat mode; cache all files in root directory (dangerous!)
     --static        static mode; assume files never change
     --offline       offline mode; never connect to server
     --limit RATE    limit download rate at a fixed K/s
     --daemon LOG    route output to log and detach
     --debug         switch from gather to debug output module''' % locals()

for _arg in _args:

  if _arg in ( '-h', '--help' ):
    sys.exit( USAGE )
  elif _arg in ( '-p', '--port' ):
    try:
      PORT = int( _args.next() )
      assert PORT > 0
    except:
      sys.exit( 'Error: %s requires a positive numerical argument' % _arg )
  elif _arg in ( '-r', '--root' ):
    try:
      ROOT = os.path.realpath( _args.next() ) + os.sep
      assert os.path.isdir( ROOT )
    except StopIteration:
      sys.exit( 'Error: %s requires a directory argument' % _arg )
    except:
      sys.exit( 'Error: invalid cache directory %s' % ROOT )
  elif _arg in ( '-v', '--verbose' ):
    VERBOSE += 1
  elif _arg in ( '-t', '--timeout' ):
    try:
      TIMEOUT = int( _args.next() )
      assert TIMEOUT > 0
    except:
      sys.exit( 'Error: %s requires a positive numerical argument' % _arg )
  elif _arg in ( '-6', '--ipv6' ):
    FAMILY = socket.AF_UNSPEC
  elif _arg == '--flat':
    FLAT = True
  elif _arg == '--static':
    STATIC = True
  elif _arg == '--offline':
    ONLINE = False
    STATIC = True
  elif _arg == '--limit':
    try:
      LIMIT = float( _args.next() ) * 1024
    except:
      sys.exit( 'Error: %s requires a numerical argument' % _arg )
  elif _arg == '--daemon':
    LOG = _args.next()
  elif _arg == '--debug':
    DEBUG = True
  else:
    sys.exit( 'Error: invalid option %r' % _arg )

MAXFILELEN = os.pathconf( ROOT, 'PC_NAME_MAX' ) - len(SUFFIX)
